### PR TITLE
Introduce "dyn newtype" pattern and convert `trait Database` to it

### PR DESCRIPTION
**Pull request by dpc**
Since we are moving more and more things to use dynamic dispatch, it might be worthwhile to introduce this pattern early, as I am already planning to use use it extensively in modularization work.

I've selected `Database` as a first trait to convert to give this approach a good test.

### CODE REVIEW

File: fedimint-api/src/db/mod.rs [hunk 1]
```diff
@@ -48,7 +50,13 @@ pub trait Database: Send + Sync {
     fn raw_apply_batch(&self, batch: DbBatch) -> Result<()>;
 }
 
-impl<'a> dyn Database + 'a {
+dyn_newtype_define! {
+    /// A handle to a type-erased database implementation
```
Comment (dpc) [inline fedimint-api/src/db/mod.rs:54]: Notably the macro handles corectly both docstring and `#[derive(...)]`.

Screenshot from 2022-09-23 12-22-37

**Comment by jkitman** [comment]
Do we really want to adopt this pattern everywhere or should we just be implementing this to break the module dependencies?

**Comment by dpc** [comment]
> Do we really want to adopt this pattern everywhere or should we just be implementing this to break the module dependencies?

Not sure what do you mean by everywhere. Generally I think wrapping `dyn Trait` in a newtype is nice and cleans up the code.

**Comment by elsirion** [comment]
We already use a more limited version of this pattern (using type defs) in quite a few places. I think this will be cleaner and the macro removes most of the overhead. "Everywhere" doesn't mean making a trait for each type imo, just where it makes sense (e.g. API boundaries for mocking, modularization), but doing so in a more consistent manner.

**Comment by elsirion** [comment]
We already use a more limited version of this pattern (using type defs) in quite a few places. I think this will be cleaner and the macro removes most of the overhead. "Everywhere" doesn't mean making a trait for each type imo, just where it makes sense (e.g. API boundaries for mocking, modularization), but doing so in a more consistent manner.

File: fedimint-api/src/macros.rs [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+            }
+
+        }
+
+        impl<I> From<I> for $name
+        where
+            I: $trait + Send + Sync + 'static,
+        {
+            fn from(i: I) -> Self {
+                Self($container::new(i))
```
Comment (jkitman) [inline fedimint-api/src/macros.rs:31]: Is this wrapping the database in an `Arc` so it can be cloned?  What happens for a `MemDatabase`?

**Comment by jkitman** [comment]
> wrapping dyn Trait in a newtype is nice and cleans up the code and I haven't noticed any downsides

Yeah I do think it is nicer, just not a pattern I've seen before, but I'll trust you Rust experts :)

File: fedimint-api/src/macros.rs (hunk 2 above)
Comment (dpc) [reply fedimint-api/src/macros.rs:31]: Yes. My understanding is that database's API needs to accommodate concurrent use.

Comment (elsirion) [reply fedimint-api/src/macros.rs:31]: > What happens for a `MemDatabase`?

It already is an `Arc<Mutex<BTreeMap<Vec<u8>, Vec<u8>>>>` internally. So maybe we can get rid of the `Arc` internally?

**Review by elsirion — APPROVED** [review]
Let's get the modularization started!

**Pull request merged**